components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -779,6 +1013,75 @@ protected void mapAccessToken(HttpServletRequest request, AuthenticationContext
+        /*
+        Federated tokens are added only of the authenticator configuration ShareFederatedToken is enabled and the
+        application has requested the federated token.
+         */
+        if (context.getAuthenticatorProperties() == null || !Boolean.parseBoolean(
+                context.getAuthenticatorProperties().get(OIDCAuthenticatorConstants.SHARE_FEDERATED_TOKEN_CONFIG)) ||
+                !requestedToShareFederatedToken(context)) {
+            return;
+        }
+
+        // If there is an existing list of federated tokens obtained in a previous step, utilizing the same list.
+        List<FederatedToken> federatedTokens;
+        Object federatedTokensObj = context.getProperty(OIDCAuthenticatorConstants.FEDERATED_TOKENS);
+        if (federatedTokensObj instanceof List) {
+            federatedTokens = (List<FederatedToken>) federatedTokensObj;
+        } else {
+            federatedTokens = new ArrayList<>();
+        }
+
+        String identityProviderName = context.getCurrentAuthenticator();
+        if (context.getExternalIdP() != null) {
+            identityProviderName = context.getExternalIdP().getIdPName();
+        }
+
+        FederatedToken federatedToken = new FederatedToken(identityProviderName,
+                oAuthResponse.getParam(OIDCAuthenticatorConstants.ACCESS_TOKEN));
+        federatedToken.setRefreshToken(oAuthResponse.getParam(OIDCAuthenticatorConstants.REFRESH_TOKEN));
+        federatedToken.setTokenValidityPeriod(oAuthResponse.getParam(OIDCAuthenticatorConstants.EXPIRES_IN));
+        federatedToken.setScope(oAuthResponse.getParam(OIDCAuthenticatorConstants.SCOPE));
+        federatedTokens.add(federatedToken);
+
+        context.setProperty(OIDCAuthenticatorConstants.FEDERATED_TOKENS, federatedTokens);
+    }
+
+    /**
+     * This returns the intersection of the allowed scopes defined at the IDP configuration and the requested scopes
+     * from the application side for federated toke sharing.
+     *
+     * @param allowedScope   The administrator defined scopes in the IDP configuration for federated toke sharing.
+     * @param requestedScope The application side requested scopes for federated toke sharing.
+     * @return The intersection of the allowed and the requested scopes for federated toke sharing.
+     */
+    private String validateScopeForTokenSharing(String allowedScope, String requestedScope) {
+
+        if (StringUtils.isNotBlank(allowedScope) && StringUtils.isNotBlank(requestedScope)) {
+            Set<String> allowedScopesSet = new HashSet<>(Arrays.asList(allowedScope.split(SPACE_REGEX)));
+            Set<String> requestedScopesSet = new HashSet<>(Arrays.asList(requestedScope.split(SPACE_REGEX)));
+
+            Set<String> subset = new HashSet<>(requestedScopesSet);
+            subset.retainAll(allowedScopesSet);
+
+            if (CollectionUtils.isNotEmpty(subset)) {
+                return StringUtils.join(subset, SPACE);
+            } else {
```

Comment:
do we need this `else` block since the if will return the anyway? 
No need to have the `isDebugEnabled` check since there's no string operations in the log

Reply:
Fixed 1fa1fba4c6641e23ef137cad4282ea9cf384b8f4